File: Cpp/Legacy89DiskKit.Cpp/src/n88_basic_configuration.cpp
#include "legacy89diskkit/cpp/n88_basic_configuration.hpp"

#include <stdexcept>

namespace legacy89diskkit::cpp
{
N88BasicConfiguration N88BasicConfigurationProvider::GetDefault(const DiskType disk_type)
{
    switch (disk_type)
    {
    case DiskType::TwoD:
        return N88BasicConfiguration{ 18, 1, 1, 12, 14, 3, 13, 256, 2048, 160, 0, 16 };
    case DiskType::TwoDD:
        return N88BasicConfiguration{ 40, 0, 1, 12, 14, 3, 13, 256, 4096, 160, 0, 16 };
    default:
        throw std::invalid_argument("Unsupported disk type for N88-BASIC");
    }
}
}

File: CSharp/Legacy89DiskKit.FileSystem.Infrastructure/FileSystem/Pc88/Models/N88BasicConfiguration.cs
using Legacy89DiskKit.DiskImage.Domain.Model;

namespace Legacy89DiskKit.FileSystem.Infrastructure.Pc88.Models;

public record N88BasicConfiguration(
    int SystemTrack,
    int SystemHead,
    int DirectorySector,
    int DirectorySectors,
    int FatSector,
    int FatSectors,
    int IdSector,
    int SectorSize,
    int ClusterSize,
    int TotalClusters,
    int ReservedClusters,
    int SectorsPerTrack
)
{
    public static N88BasicConfiguration GetDefault(DiskType diskType)
    {
        return diskType switch
        {
            DiskType.TwoD => new N88BasicConfiguration(
                SystemTrack: 18, 
                SystemHead: 1, 
                DirectorySector: 1, 
                DirectorySectors: 12, 
                FatSector: 14, 
                FatSectors: 3, 
                IdSector: 13, 
                SectorSize: 256, 
                ClusterSize: 2048, 
                TotalClusters: 160, // 40 tracks * 2 sides * 2 clusters/track (8 sectors per cluster)
                ReservedClusters: 0, 
                SectorsPerTrack: 16),
            DiskType.TwoDD => new N88BasicConfiguration(
                SystemTrack: 40, 
                SystemHead: 0, 
                DirectorySector: 1, 
                DirectorySectors: 12, 
                FatSector: 14, 
                FatSectors: 3, 
                IdSector: 13, 
                SectorSize: 256, 
                ClusterSize: 4096, 
                TotalClusters: 160, 
                ReservedClusters: 0, 
                SectorsPerTrack: 16),
            _ => throw new NotSupportedException($"DiskType {diskType} is not supported for N88-BASIC yet.")
        };
    }
}
